src/main/java/bio/terra/workspace/db/DataReferenceDao.java

```diff
@@ -0,0 +1,142 @@
+package bio.terra.workspace.db;
+
+import bio.terra.workspace.app.configuration.WorkspaceManagerJdbcConfiguration;
+import bio.terra.workspace.common.exception.ValidationException;
+import bio.terra.workspace.generated.model.DataReferenceDescription;
+import bio.terra.workspace.generated.model.DataReferenceDescription.CloningInstructionsEnum;
+import bio.terra.workspace.generated.model.DataReferenceDescription.ReferenceTypeEnum;
+import bio.terra.workspace.generated.model.DataReferenceList;
+import bio.terra.workspace.generated.model.ResourceDescription;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.stereotype.Component;
+
+@Component
+public class DataReferenceDao {
+
+  private final NamedParameterJdbcTemplate jdbcTemplate;
+
+  @Autowired
+  public DataReferenceDao(WorkspaceManagerJdbcConfiguration jdbcConfiguration) {
+    this.jdbcTemplate = new NamedParameterJdbcTemplate(jdbcConfiguration.getDataSource());
+  }
+
+  public DataReferenceList enumerateDataReferences(
+      String workspaceId, String owner, int offset, int limit, String filterControlled) {
+    List<String> whereClauses = new ArrayList<>();
+    whereClauses.add("(ref.workspace_id = :id)");
+    whereClauses.add(filterControlledWhereClause(filterControlled, "ref"));
+    whereClauses.add(uncontrolledOrVisibleResourcesClause(owner, "resource", "ref"));
+    String filterSql = combineWhereClauses(whereClauses);
+    String sql =
+        "SELECT ref.workspace_id, ref.reference_id, ref.name, ref.resource_id, ref.credential_id, ref.cloning_instructions, ref.reference_type, ref.reference, "
+            + "resource.resource_id, resource.associated_app, resource.is_visible, resource.owner, resource.attributes  "
+            + "FROM workspace_data_reference AS ref "
+            + "LEFT JOIN workspace_resource AS resource ON ref.resource_id = resource.resource_id "
+            + filterSql
+            + " ORDER BY ref.reference_id "
+            + "OFFSET :offset "
+            + "LIMIT :limit";
```

Comment:
nit: I'd suggest being consistent about where the space separators are in the string. I put them at the start after a + because it is easier for me to forget to put them at the end.

Reply:
Done, also added a space to the start of the WHERE clause to maintain the pattern